pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -137,6 +162,109 @@ func (grb *globalRoleBindingLifecycle) deleteAdminBinding(obj *v3.GlobalRoleBind
+			_, err = grb.crtbClient.Create(&v3.ClusterRoleTemplateBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "crtb-grb-",
+					Namespace:    cluster.Name,
+					// the owner ref needs to be mutable by the k8s garbage collector but we need
+					// a way to identify what CRTBs are from GRBs unambiguously for validation
+					Labels: map[string]string{
+						grbOwnerLabel: globalRoleBinding.Name,
+					},
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: v3.GlobalRoleBindingGroupVersionKind.GroupVersion().String(),
+							Kind:       v3.GlobalRoleBindingGroupVersionKind.Kind,
+							Name:       globalRoleBinding.Name,
+							UID:        globalRoleBinding.UID,
+						},
+					},
+				},
+				ClusterName:        cluster.Name,
+				RoleTemplateName:   wantRT,
+				UserName:           globalRoleBinding.UserName,
+				GroupPrincipalName: globalRoleBinding.GroupPrincipalName,
+			})
+			// we don't immediately return so that we can create as many CRTBs as we can
+			if err != nil {
+				logrus.Errorf("failed to create crtb for globalRoleBinding %s in cluster %s: %s", globalRoleBinding.Name, cluster.Name, err.Error())
+				missedClusters = true
+			}
+		}
+	}
+	if missedClusters {
+		return fmt.Errorf("unable to create backing crtbs for globalRoleBinding %s, some permissions may be missing", globalRoleBinding.Name)
+	}
+	return nil
+}
+
+// purgeCorruptRoles removes any CRTBs which were created for this role in the past, but are no longer valid. Returns
+// a slice of the roles that were in wantRTs but are not yet covered by the existing CRTBs
+func (grb *globalRoleBindingLifecycle) purgeCorruptRoles(wantRTs []string, cluster *v3.Cluster, binding *v3.GlobalRoleBinding) ([]string, error) {
+	currentRTs := map[string]struct{}{}
+	for _, wantRT := range wantRTs {
+		currentRTs[wantRT] = struct{}{}
+	}
+	currentCRTBs, err := grb.crtbCache.GetByIndex(crtbGrbOwnerIndex, fmt.Sprintf("%s/%s", cluster.Name, binding.Name))
+	if err != nil {
+		return nil, fmt.Errorf("unable to get CRTBs for cluster %s: %w", cluster.Name, err)
+	}
+	for _, crtb := range currentCRTBs {
+		_, rtOk := currentRTs[crtb.RoleTemplateName]
+		if rtOk &&
+			crtb.ClusterName == cluster.Name &&
+			crtb.UserName == binding.UserName &&
+			crtb.GroupPrincipalName == binding.GroupPrincipalName &&
+			crtb.DeletionTimestamp == nil {
+			delete(currentRTs, crtb.RoleTemplateName)
+		} else {
+			// if there's a CRTB to a roleTemplate not currently specified, or to a different cluster/subject, remove it
+			// CRTBs can't update some of these fields, so the safest method is to delete/re-create
+			err := grb.crtbClient.DeleteNamespaced(crtb.Namespace, crtb.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("unable to delete backing crtb %s for globalRoleBinding %s: %w", crtb.Name, binding.Name, err)
+			}
```

Comment:
There may be other CRTBs that need to be deleted, if one is stuck `terminating` (or failed for some other reason), the subsequent CRTBs will have to wait until the next reconciliation. What are your thoughts on not returning here, but rather keep track of all CRTB names and bindings that errored, and return the error _after_ the `for` loop has finished?

Reply:
I've changed this to now try and delete everything. In addition, deleting is no longer a failure condition - we will continue to sync valid roles even if we can't delete the invalid roles.